Fix error handling in CMS_EncryptedData_encrypt

bernd-edlinger opened the pull request:
That caused several memory leaks in case of error. Also when the CMS object that is created by CMS_EncryptedData_encrypt is not used in the normal way, but instead just deleted by CMS_ContentInfo_free some memory was lost.

Fixes #21985

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [x] tests are added or updated

t8m reviewed (approved):
@bernd-edlinger If you promise making patches against newer branches I am OK merging this to 1.1.1 as this looks like a potential security hardening.

bernd-edlinger commented:
OK, deal :)

t8m commented:
We need a review from committers ASAP if this should go in before the 1.1.1 EOL

beldmit reviewed (approved):
LGTM

bernd-edlinger commented on crypto/cms/cms_sd.c:
```diff
@@ -745,8 +749,13 @@ int CMS_SignerInfo_verify(CMS_SignerInfo *si)
+    if (si->pctx != NULL) {
+        EVP_PKEY_CTX_free(si->pctx);
+        si->pctx = NULL;
+    }
```
While working out a PR for master, I realised that this function
can be called twice if e.g. CMS_verify is called twice, and that would
create a memory leak.

bernd-edlinger commented:
@t8m @hlandau @beldmit please re-confirm.

beldmit commented:
My approval persists.

bernd-edlinger commented:
Sorry, I had to update the PR once more, I added two fuzz/corpora/cms files, one
directly from the failed CI workflow in master which turned out to directly affect this PR too.
I believe I have now fixed the issue correctly, and just-in-time for the upcoming 1.1.1 release party :-)
So please everyone re-onfirm your approvals.
Thanks!

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

bernd-edlinger commented:
Please re-approve.
Thanks!

beldmit reviewed (approved):
LGTM

t8m commented:
Merged to 1.1.1 branch. Thank you for your contribution.
